controllers/tykoasapidefinition_controller.go

```diff
@@ -188,6 +220,173 @@ func (r *TykOasApiDefinitionReconciler) Reconcile(ctx context.Context, req ctrl.
+			}
+
+			return nil
+		}
+
+		var baseOasAPI v1alpha1.TykOasApiDefinition
+
+		if err := r.Get(ctx, types.NamespacedName{
+			Name:      versionedTykOAS.Status.BaseVersionName,
+			Namespace: versionedTykOAS.Status.BaseVersionNamespace,
+		}, &baseOasAPI); err != nil {
+			if k8serrors.IsNotFound(err) {
+				log.Info("failed to get TykOasApiDefinition CR", "err", err.Error())
+			}
+
+			return err
+		}
+
+		errMsg := "Cannot delete API. update the Base OasAPI to delete"
+
+		versionedTykOAS.Status.LatestTransaction = v1alpha1.TransactionInfo{
+			Time:   metav1.Now(),
+			Status: v1alpha1.TransactionStatus(errMsg),
+		}
+
+		if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+			return err
+		}
+	}
+
+	// this cleans up dangling versioned OAS APIs
+	if len(versionedTykOAS.Status.BaseVersionName) > 0 {
+		var baseOasAPI v1alpha1.TykOasApiDefinition
+
+		ns := versionedTykOAS.Status.BaseVersionNamespace
+		if ns == "" {
+			ns = "default"
+		}
+
+		if err := r.Get(ctx, types.NamespacedName{
+			Name:      versionedTykOAS.Status.BaseVersionName,
+			Namespace: ns,
+		}, &baseOasAPI); err != nil {
+			if k8serrors.IsNotFound(err) {
+				log.Info("base oas version not found, detaching version from the base OAS API.")
+
+				versionedTykOAS.RemoveOASVersionStatus()
+
+				if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+					return err
+				}
+
+				return nil
+			}
+
+			return err
+		}
+
+		var oasAPIPresent bool
+
+		for _, v := range baseOasAPI.Spec.Versioning.Versions {
+			if v.OasApiDefinitionRef == versionedTykOAS.Name {
+				oasAPIPresent = true
+			}
+		}
+
+		if !oasAPIPresent {
+			versionedTykOAS.RemoveOASVersionStatus()
+
+			if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	return nil
+}
+
+func (r *TykOasApiDefinitionReconciler) handleVersioningTykOASAPI(ctx context.Context,
+	tykBaseOASCrd *v1alpha1.TykOasApiDefinition,
+	log logr.Logger,
+	tykBaseOASDoc,
+	baseAPIID string,
+) error {
+	var versions []*TykOasApiDefinitionVersion
+
+	for _, v := range tykBaseOASCrd.Spec.Versioning.Versions {
+		var versionOasAPI v1alpha1.TykOasApiDefinition
+
+		if err := r.Get(ctx, types.NamespacedName{
+			Name:      v.OasApiDefinitionRef,
+			Namespace: v.Namespace,
+		}, &versionOasAPI); err != nil {
+			if k8serrors.IsNotFound(err) {
+				log.Info("failed to get versioned TykOasApiDefinition CR", "err", err.Error())
+			}
+
+			return err
+		}
+
+		if tykBaseOASCrd.Spec.Versioning.Default == v.Name {
+			versionOasAPI.Status.IsDefaultVersion = true
+		}
+
+		versionOasAPI.Status.BaseVersionName = tykBaseOASCrd.Name
+		versionOasAPI.Status.BaseVersionNamespace = tykBaseOASCrd.Namespace
+		versionOasAPI.Status.IsVersionedAPI = true
+
+		version := TykOasApiDefinitionVersion{
+			ID:   versionOasAPI.Status.ID,
+			Name: v.Name,
+		}
+
+		versions = append(versions, &version)
+
+		if err := r.Status().Update(ctx, &versionOasAPI); err != nil {
```

Comment:
updating status here is tricky because this method is called within CreateOrUpdate which reconciles resources within its closure function. and updating the status would cause conflicts on k8s side - which might lead redundant reconciliations to recover.

Reply:
This type of update needs to happen when the baseAPI is marked for deletion. What is happening in the function is that, all versions are being detached from the base API when a delete of the base API is triggered.

Reply:
i see why it is needed. the problem is related how kubernetes api handles updates. this function called inside CreateOrUpdate, [here](https://github.com/TykTechnologies/tyk-operator/blob/153761130cf0fbb95a9a2f3b3b010b57eea9203a/controllers/tykoasapidefinition_controller.go#L183).
If you check [the docs](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.0/pkg/controller/controllerutil#CreateOrUpdate) of CreateOrUpdate, you will see that the CreateOrUpdate function fetches the resource from kubernetes, then runs the function that we passed into that and Updates the resource. In the meantime, if you update resource's Status field, there might be a drift in `resourceVersion` of the resource, which will cause conflicts on kubernetes side.
Let me try to simplify, assume that you have a CRD called `X` which has `resourceVersion: 1`.
```
function CreateOrUpdate starts
        fetches X from API server
        sees X is having a resourceVersion: 1
        runs mutate function that we passed into CreateOrUpdate
                ---> mutate function starts
                the mutate function updates status which will update the resourceVersion 
                    of the object and new resourceVersion will be different than 1.
               ---> mutate function finishes
        CreateOrUpdate sends the object to the Api Server to update it.
        It will fail because the object that we reconciled within API Server 
              has resourceVersion: 1 while the API server has the 
              same object with different resourceVersion. 
              That's why the manager will requeue the object again 
              and this loop might stuck. 
```

Reply:
If i get you correctly, in this function the x will be the baseAPI. However, it is not the baseAPI that is being updated, but the versions and this update happens only when the base API has been marked for deletion.

Reply:
yes, you are right - mb. please ignore the comment